### DISCUSSION TITLE
OR-Tools non-present optional interval variables leading to duplicate solutions

### DISCUSSION THREAD

**Discussion by raunaqsingh2020** [google/or-tools#3606] [CP-SAT questions]
**What version of OR-Tools and what language are you using?**
Version: v9.5
Language: Python3.10.8

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
CP-SAT

**What operating system (Linux, Windows, ...) and version?**
macOS Monterey 12.6

**What did you do?**
I am currently working on a scheduling problem (creating schedules of class sections for students). In order to do this, I had boolean variables for each course section (corresponding to whether or not the section would be in the resulting schedule), and (optional) interval variables for the section meetings. These interval variables are present when their corresponding section variable is True (as I only care about avoiding overlapping between sections that are actually included in the schedule). I wanted to see all possible solutions using the CPSolver parameter  `enumerate_all_solutions`.

Here are some snippets of the code:

```py
for section in course.sections.all():
	sections[section.full_code] = model.NewBoolVar('section_%s' % (section.full_code))

	for i, meeting in enumerate(section.meetings.all()):
		start, duration = meeting_to_interval(meeting)
                meeting_interval = model.NewOptionalFixedSizeIntervalVar(start, duration, sections[section.full_code], 'section_%s_interval_%d' % (section.full_code, i))
                all_section_intervals.append(meeting_interval)

...

model.AddNoOverlap(all_section_intervals)
```

```py
solver.parameters.enumerate_all_solutions = True

class SolutionHandler(cp_model.CpSolverSolutionCallback):
	def __init__(self, schedules, sections, limit):
		cp_model.CpSolverSolutionCallback.__init__(self)
		self.__schedules = schedules
   	        self.__sections = sections
		self.__solution_count = 0
		self.__solution_limit = limit

        def on_solution_callback(self):
		self.__solution_count += 1
		recommended_schedule = [section for section in self.__sections if self.Value(self.__sections[section])]
		self.__schedules += [recommended_schedule]

		if self.__solution_count >= self.__solution_limit:
			self.StopSearch()
        
	def solution_count(self):
		return self.__solution_count
```

**What did you expect to see**

I expected to see unique solutions (lists of sections).

**What did you see instead?**

I got a number of duplicate solutions. For example, when I limited the number of solutions to 5 and locked in some sections, the results were:
```
['CIS-5710-401', 'CIS-5050-001', 'GUJR-0400-680', 'CIS-5590-001', 'CIS-5590-201']
['CIS-5710-401', 'CIS-5050-001', 'RUSS-1200-401', 'CIS-5590-001', 'CIS-5590-201']
['CIS-5710-401', 'CIS-5050-001', 'PUNJ-0400-680', 'CIS-5590-001', 'CIS-5590-201']
['CIS-5710-401', 'CIS-5050-001', 'GUJR-0400-680', 'CIS-5590-001', 'CIS-5590-201']
['CIS-5710-401', 'CIS-5050-001', 'GUJR-0400-680', 'CIS-5590-001', 'CIS-5590-201']
```

As you can see, 3 out of the 5 schedules are the exact same. If I remove the limit entirely, the solver will repeatedly list out the 3 distinct schedules above. I believe this is because of the `NewOptionalFixedSizeIntervalVar`'s (as when I remove them and the NoOverlap constraint, the resulting solutions are all distinct). I think the solver is interpreting different values of these interval variables to mean distinct solutions, but I am not sure why / if that is intended. They are directly connected to the boolean variables (i.e. present based on the boolean variables), so there shouldn't be any difference between solutions with the same values for those boolean variables. 

E.g. for two solutions where the result is `['CIS-5710-401', 'CIS-5050-001', 'GUJR-0400-680', 'CIS-5590-001', 'CIS-5590-201']`, my solver would have that `solver.Value(sections['CIS-5710-401']) == 1`, `solver.Value(sections['CIS-5050-001']) == 1`, etc. but somehow (I believe) be considered different because of the interval variables.

**Anything else we should know about your project / environment**
Nope, thanks!

**Comment by lperron** [comment]
Intervals are fixed ? If not, there could be different solutions with different start values.
If yes, you do not need intervals, although I understand this is nice.

Do you have Boolean variables that do not appear in a interval ?

**Comment by raunaqsingh2020** [comment]
The intervals are fixed, but I'm not sure how I could avoid using them. They aren't very simple, in that I won't be able to simply make buckets of sections for each interval. E.g. there may be a section from 1pm-2:30pm, and another section from 1:30pm-2:15pm. How would you recommend approaching that situation without interval variables? Also, does it make sense then that there are repeated solutions?

I'm not entirely sure what you mean by boolean variables that do not appear in an interval. I have boolean variables for courses and boolean variables for the corresponding sections of those courses. A section variable is true if and only if the course variable is true. I confirmed that these course variables aren't causing different solutions though.

**Comment by lperron** [comment]
If both Boolean vars are equivalent, just create one.

**Comment by raunaqsingh2020** [comment]
They aren't equivalent. A course has multiple sections, and the course variable is an indicator that *any* of those sections is true.

**Comment by raunaqsingh2020** [comment]
@lperron I am still unsure of why this isn't a bug / issue that needs fixing. If there are some optional fixed interval variables that are only present based off some boolean variables, how can they result in different solutions for the same values of those boolean variables? Thanks!

**Comment by lperron** [comment]
This code is rather well tested.

And unless you can send me a small reproducible model, there is nothing I
can do.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00



Le mar. 27 déc. 2022 à 10:47, Raunaq Singh ***@***.***> a
écrit :
